src/cmd-koji-upload

```diff
@@ -0,0 +1,773 @@
+                    level=sl)
+
+
+def kinit(keytab, principle):
+    """
+    Execute the Kerberos Auth using the provided keytab:
+    :param keytab: file name of the keytab used for Kerberos Authentication
+    :type str
+    :param principle: name of the keytab's principle
+    :type str
+    """
+    if keytab is None:
+        raise Exception("keytab file is not defined")
+
+    log.info("using kerberos auth via %s", keytab)
+    try:
+        _ = subprocess.check_output([
+            "kinit", "-f", "-t", keytab, "-k", principle])
+        klist_out = subprocess.check_output(["klist", "-a"])
+        log.debug("authenticated: \n%s", klist_out.decode("utf-8"))
+    except Exception as err:
+        raise Exception("failed to auth: ", err)
+
+
+def get_koji_fileinfo(fname):
+    """
+    Get archive_info is a helper to get the content generator parts
+    :param fname: file name to check for Koji content type
+    :type str
+    :returns str, str, str, str
+    """
+    ext = (os.path.splitext(fname)[1]).replace(".", "")
+    if ext == "gz":
+        ext = fname.split(".")[-2]
+
+    _, force_extension = special_name(fname)
+    if force_extension is not None:
+        ext = force_extension
+
+    try:
+        return KOJI_CG_TYPES[ext]
+    except KeyError:
+        log.debug("file type %s was not found in lookup", ext)
+        return None, None, None, None
+
+
+def special_name(fname):
+    """
+    Helper for handling special files such as unlabeled raw files. Files
+    that need have their names mangled should be handled here. Koji has strict
+    rules that ensure the file-type to match a specific extension.
+
+    :param fname: name of file to check if it needs to be renamed
+    :type str
+    :returns str, bool
+    """
+
+    # koji doesn't understand raw.gz, so swap 'em
+    # request filed with Koji maintainers to fix this.
+    if fname.endswith("raw.gz"):
+        return fname.replace("raw.gz", "gz.raw"), "raw"
+
+    # handle special files such as installers kernels.
+    for ending in RENAME_RAW:
+        if fname.endswith(ending):
+            return "%s.raw" % fname, "raw"
+
+    return fname, None
+
+
+def koji_upload(fname):
+    """
+    Helper to return if a file should be uploaded
+    :param fname: name of file to check against Koji table for uploading.
+    :type str
+    :returns bool
+    """
+    found, _, _, _ = get_koji_fileinfo(fname)
+    if found:
+        return True
+    return False
+
+
+def md5sum_file(path):
+    """
+    Calculates the md5 sum from a path.
+    :param path: file name to checksum
+    :returns str
+    """
+    h = hashlib.md5()
+    with open(path, 'rb', buffering=0) as data:
+        for b in iter(lambda: data.read(128 * 1024), b''):
+            h.update(b)
+    return h.hexdigest()
+
+
+def load_json(path):
+    """
+    Shortcut for loading json from a file path.
+    This is cmdlib, but cmdlib is py3 and we have to use py2.
```

Comment:
The problem isn't with Koji supporting Python3, its that the Kerberos Auth in Python3 does work against the internal Koji for RedHat
```
[Tue Apr 30 21:48:45.716187 2019] [auth_gssapi:error] [pid 5706] [client 10.0.65.14:52936] NO AUTH DATA Client did not send any authentication headers
[Tue Apr 30 21:48:46.168817 2019] [auth_gssapi:error] [pid 5706] [client 10.0.65.14:52936] GSS ERROR gss_localname() failed: [The operation or option is not available or unsupported (No such file or directory)]```

After inquiring with the Koji team, I was told that using Py2 w/ the `python2-krbv`  is the fix. I haven't found a clear bug report on this specific error.

Reply:
I think it would be really useful to include that comment here or at the top of the script so we can know why the choice was made and also move to py3 in the future when it's no longer a problem.

Reply:
Ack, good call. Added.